File: src/infinicore/ops/tbmv/tbmv.cc
#include "infinicore/ops/tbmv.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Tbmv);

Tbmv::Tbmv(const Tensor &a, Tensor x, int uplo, int trans, int diag, size_t k) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, x);
    INFINICORE_GRAPH_OP_DISPATCH(x->device().getType(), a, x, uplo, trans, diag, k);
}

void Tbmv::execute(const Tensor &a, Tensor x, int uplo, int trans, int diag, size_t k) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Tbmv, a, x, uplo, trans, diag, k);
}

void tbmv_(const Tensor &a, Tensor x, int uplo, int trans, int diag, size_t k) {
    Tbmv::execute(a, x, uplo, trans, diag, k);
}

} // namespace infinicore::op

File: python/infinicore/ops/tbmv.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def tbmv(
    a: Tensor,
    x: Tensor,
    *,
    uplo: int = 0,
    trans: int = 0,
    diag: int = 0,
    k: int,
):
    _infinicore.tbmv_(
        a._underlying,
        x._underlying,
        uplo,
        trans,
        diag,
        k,
    )

    return x
